src/gfn/utils/prob_calculations.py

```diff
@@ -0,0 +1,232 @@
+
+def get_traj_pfs_and_pbs(
+    pf: GFNModule,
+    pb: GFNModule,
+    trajectories: Trajectories,
+    fill_value: float = 0.0,
+    recalculate_all_logprobs: bool = False,
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    # fill value is the value used for invalid states (sink state usually)
+    if trajectories.is_backward:
+        raise ValueError("Backward trajectories are not supported")
+
+    # uncomment next line for debugging
+    # assert trajectories.states.is_sink_state[:-1].equal(trajectories.actions.is_dummy)
+
+    log_pf_trajectories = get_traj_pfs(
+        pf,
+        trajectories,
+        fill_value=fill_value,
+        recalculate_all_logprobs=recalculate_all_logprobs,
+    )
+    log_pb_trajectories = get_traj_pbs(pb, trajectories, fill_value=fill_value)
+
+    assert log_pf_trajectories.shape == (
+        trajectories.max_length,
+        trajectories.n_trajectories,
+    )
+    assert log_pb_trajectories.shape == (
+        trajectories.max_length,
+        trajectories.n_trajectories,
+    )
+
+    return log_pf_trajectories, log_pb_trajectories
+
+
+def get_traj_pfs(
+    pf: GFNModule,
+    trajectories: Trajectories,
+    fill_value: float = 0.0,
+    recalculate_all_logprobs: bool = False,
+) -> torch.Tensor:
+    state_mask = ~trajectories.states.is_sink_state
+    action_mask = ~trajectories.actions.is_dummy
+
+    valid_states = trajectories.states[state_mask]
+    valid_actions = trajectories.actions[action_mask]
+
+    if valid_states.batch_shape != tuple(valid_actions.batch_shape):
+        raise AssertionError("Something wrong happening with log_pf evaluations")
+
+    if has_log_probs(trajectories) and not recalculate_all_logprobs:
+        log_pf_trajectories = trajectories.log_probs
+    else:
+        if trajectories.estimator_outputs is not None and not recalculate_all_logprobs:
+            estimator_outputs = trajectories.estimator_outputs[action_mask]
+        else:
+            masked_cond = None
+            if trajectories.conditioning is not None:
+                cond_dim = (-1,) * len(trajectories.conditioning.shape)
+                traj_len = trajectories.states.tensor.shape[0]
+                masked_cond = trajectories.conditioning.unsqueeze(0).expand(
+                    (traj_len,) + cond_dim
+                )[state_mask]
+
+            estimator_outputs = check_cond_forward(pf, "pf", valid_states, masked_cond)
+
+        # Calculates the log PF of the actions sampled off policy.
+        valid_log_pf_actions = pf.to_probability_distribution(
+            valid_states, estimator_outputs
+        ).log_prob(
+            valid_actions.tensor
+        )  # Using the actions sampled off-policy.
+
+        log_pf_trajectories = torch.full_like(
+            trajectories.actions.tensor[..., 0],
+            fill_value=fill_value,
+            dtype=torch.float,
+        )
+        log_pf_trajectories[action_mask] = valid_log_pf_actions
+
+    return log_pf_trajectories
+
+
+def get_traj_pbs(
```

Comment:
For consistency, I'd also suggest we use here `trajectory` (ditto for all other instances)